### ISSUE TITLE
Verification

### ISSUE THREAD

**Issue by mogronalol** [SpectoLabs/hoverfly-java#42]
The ability to verify requests which have been made would be useful, in a way similar to Mockito but for http. The problem stems around us being dependant on the functionality to be implemented in Hoverfly itself, although it is currently on the roadmap.

**Comment by Khazrak** [comment]
I agree, verification would be awesome. Both that the request have been made and how many times (so we can test that it doesn't call it multiple times)

**Comment by JohnFDavenport** [comment]
@Khazrak that's an interesting point on multiple calls. I like use-cases. Here's mine. I've seen problems when using a third-party service provider has per transaction costs. If developers are not alert they may make multiple service calls in a longer flow. If that goes undetected the costs can be significant. However, my use-case is for integration or end-to-end testing.

**Comment by Khazrak** [comment]
Some services I worked on with a Mocked services to talk to, some times our service hid retries or a old method did several calls. It got caught after they used Wiremock and verify. It's a good check that they did call it and only called it once.

Another function that would be awesome is to record 2 calls to same url with different results. 
Use-case of that is:
1. Call "list containers" -> assert size
2. Call "remove container X"
3. Call "list containers" -> assert size 1 less then before

**Comment by tommysitu** [comment]
Verify the number of requests for exactly / atLeast / atMost n times are useful especially your http client has retry logic. 

Another thing is you can verify specific header is sent. 

What you described last is stunning stateful behaviour which I believe it is in the road map. It might be useful for selenium testing.

**Comment by mogronalol** [comment]
Here is the upstream issue blocking this https://github.com/SpectoLabs/hoverfly/issues/501#issuecomment-293191322

**Comment by tommysitu** [comment]
Just sketching out how we can do verification: 
```
// The method signatures
HoverflyDsl.verify(RequestMatcherBuilder builder, HoverflyVerific assertions)
HoverflyDsl.verifyZeroRequest(StubServiceBuilder builder)

// Examples
HoverflyDsl.verify(service(matches("www.*-test.com")).get("/api/bookings/1"), times(2));
HoverflyDsl.verifyZeroRequest(service("cloud.api.com"));

```

`HoverflyAssertions` can support these static methods, similar to Mockito:
```
never()
times(int count)
atLeast(int count)
atMost(int count)
atLeastOnce()
```

**Comment by mogronalol** [comment]
You can probably overload the method also, so something like:

```
HoverflyDsl.verify(requestMatcherBuilder) // Defaults to once
```

I'm not sure about `verifyZeroRequest` to a base URL. Because you could just do:

```
HoverflyDsl.verify(matcherWithOnlyBaseUrl, never()) // Defaults to once
```

**Comment by tommysitu** [comment]
Yeah, that's true. But I am thinking the `HoverflyDsl.verifyZeroRequest(StubServiceBuilder)` is a quick way to verify the test is not calling a particular service at all. Its argument only takes a value such as `service("api.com")` not the full RequestMatcherBuilder. 

Use case would be I want to quickly verify the test won't trigger a call to my notification microservice.

**Comment by mogronalol** [comment]
I guess internally it could call

```
HoverflyDsl.verifyZeroCallsToService(baseUrl) ->HoverflyDsl.verify(service(baseUrl), never())

**Comment by tommysitu** [comment]
yeah, that will do

**Comment by tommysitu** [comment]
Implemented in release 0.7.0
